File: crates/beamr/tests/fixtures/is_function_bif.erl
%% Fixture exercising erlang:is_function/1 and is_function/2 as callable
%% BIFs rather than the literal-arity is_function2 test opcode: the
%% compiler emits the guard-BIF instruction (`bif` with an
%% erlang:is_function import) for body-position calls and for guards whose
%% arity is not a compile-time literal. Compile with
%% `erlc is_function_bif.erl` (OTP 25+) and commit the .beam next to this
%% source.
-module(is_function_bif).
-export([body_one/0, body_two/0, guard_variable_arity/0, guard_badarg/0,
         arity_badarg/0, id/1, check/2]).

%% Exported so the compiler cannot constant-fold the type tests or infer
%% argument types at the call sites (type inference would otherwise turn
%% the variable-arity guard back into the is_function2 test opcode).
id(X) -> X.

%% Body-position is_function/1 on a fun and a non-fun.
body_one() ->
    {is_function(id(fun() -> ok end)), is_function(id(nope))}.

%% Body-position is_function/2: exact arity matches, others do not.
body_two() ->
    F = id(fun(A, B) -> {A, B} end),
    {is_function(F, 2), is_function(F, 3), is_function(id(nope), 2)}.

%% A variable arity compiles to the erlang:is_function/2 guard BIF.
guard_variable_arity() ->
    F = id(fun(A) -> A end),
    {check(F, 1), check(F, 2), check(nope, 1)}.

check(F, N) when is_function(F, N) -> matched;
check(_, _) -> fallthrough.

%% An invalid arity (negative or non-integer) in GUARD position must fail
%% the guard and take the false branch, not crash the process.
guard_badarg() ->
    F = id(fun(A) -> A end),
    {check(F, -1), check(F, not_an_arity)}.

%% A negative arity in body position raises badarg.
arity_badarg() ->
    is_function(id(fun() -> ok end), id(-1)).

File: crates/beamr/tests/is_function_bif.rs
//! Regression test for erlang:is_function/1 and is_function/2 as callable
//! BIFs.
//!
//! Only the literal-arity `is_function2` test OPCODE was handled; no
//! `is_function` BIF was registered anywhere, so imports of
//! erlang:is_function resolved to `Deferred` (erlang is never a loaded
//! module) and every body-position call or variable-arity guard — which
//! the compiler emits as the guard-BIF instruction with an
//! erlang:is_function import — crashed the process at call time.
//!
//! Fixture source: `tests/fixtures/is_function_bif.erl`.

use std::sync::Arc;

use beamr::atom::AtomTable;
use beamr::ets::copy::OwnedTerm;
use beamr::loader::load_module;
use beamr::module::ModuleRegistry;
use beamr::native::{
    BifRegistryImpl, bifs::register_gate1_bifs, gate3_bifs::register_gate3_bifs,
    process_bifs::register_gate2_bifs, stdlib_stubs::register_stdlib_stubs,
};
use beamr::process::ExitReason;
use beamr::scheduler::{Scheduler, SchedulerConfig};
use beamr::term::Term;
use beamr::term::boxed::Tuple;

fn start_scheduler(atoms: &Arc<AtomTable>) -> Scheduler {
    let bifs = BifRegistryImpl::new();
    register_gate1_bifs(&bifs, atoms).expect("gate 1 bifs");
    register_gate2_bifs(&bifs, atoms).expect("gate 2 bifs");
    register_gate3_bifs(&bifs, atoms).expect("gate 3 bifs");
    register_stdlib_stubs(&bifs, atoms).expect("stdlib stubs");
    let registry = Arc::new(ModuleRegistry::new());
    let (_module, _report) = load_module(
        include_bytes!("fixtures/is_function_bif.beam"),
        atoms,
        &registry,
        &bifs,
    )
    .expect("is_function_bif fixture loads");
    Scheduler::with_code_server(
        SchedulerConfig {
            thread_count: Some(1),
            ..SchedulerConfig::default()
        },
        registry,
        Arc::clone(atoms),
        Arc::new(bifs),
    )
    .expect("scheduler starts")
}

/// Runs `is_function_bif:<function>/0` to completion.
fn call(scheduler: &Scheduler, atoms: &AtomTable, function: &str) -> (ExitReason, OwnedTerm) {
    let module = atoms.intern("is_function_bif");
    let function = atoms.intern(function);
    let pid = scheduler
        .spawn(module, function, Vec::new())
        .expect("spawn fixture function");
    scheduler.run_until_exit(pid)
}

fn assert_atom_tuple(atoms: &AtomTable, term: Term, expected: &[&str], context: &str) {
    let tuple = Tuple::new(term).unwrap_or_else(|| panic!("{context}: expected tuple result"));
    assert_eq!(tuple.arity(), expected.len(), "{context}: tuple arity");
    for (index, name) in expected.iter().enumerate() {
        let atom = atoms.intern(name);
        assert_eq!(
            tuple.get(index),
            Some(Term::atom(atom)),
            "{context}: element {index} should be {name}"
        );
    }
}

#[test]
fn body_position_is_function_1_returns_booleans() {
    let atoms = Arc::new(AtomTable::with_common_atoms());
    let scheduler = start_scheduler(&atoms);
    let (reason, result) = call(&scheduler, &atoms, "body_one");
    assert_eq!(reason, ExitReason::Normal);
    assert_atom_tuple(&atoms, result.root(), &["true", "false"], "body_one");
    scheduler.shutdown();
}

#[test]
fn body_position_is_function_2_checks_exact_arity() {
    let atoms = Arc::new(AtomTable::with_common_atoms());
    let scheduler = start_scheduler(&atoms);
    let (reason, result) = call(&scheduler, &atoms, "body_two");
    assert_eq!(reason, ExitReason::Normal);
    assert_atom_tuple(
        &atoms,
        result.root(),
        &["true", "false", "false"],
        "body_two",
    );
    scheduler.shutdown();
}

#[test]
fn variable_arity_guard_routes_through_the_bif() {
    let atoms = Arc::new(AtomTable::with_common_atoms());
    let scheduler = start_scheduler(&atoms);
    let (reason, result) = call(&scheduler, &atoms, "guard_variable_arity");
    assert_eq!(reason, ExitReason::Normal);
    assert_atom_tuple(
        &atoms,
        result.root(),
        &["matched", "fallthrough", "fallthrough"],
        "guard_variable_arity",
    );
    scheduler.shutdown();
}

#[test]
fn negative_arity_in_body_position_is_badarg() {
    let atoms = Arc::new(AtomTable::with_common_atoms());
    let scheduler = start_scheduler(&atoms);
    let (reason, _result) = call(&scheduler, &atoms, "arity_badarg");
    assert_eq!(reason, ExitReason::Error);
    scheduler.shutdown();
}
